src/serializers/type_serializers/union.rs

```diff
@@ -198,24 +202,180 @@ impl TypeSerializer for UnionSerializer {
+impl BuildSerializer for TaggedUnionSerializer {
     const EXPECTED_TYPE: &'static str = "tagged-union";
 
     fn build(
         schema: &Bound<'_, PyDict>,
         config: Option<&Bound<'_, PyDict>>,
         definitions: &mut DefinitionsBuilder<CombinedSerializer>,
     ) -> PyResult<CombinedSerializer> {
-        let schema_choices: Bound<'_, PyDict> = schema.get_as_req(intern!(schema.py(), "choices"))?;
-        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(schema_choices.len());
+        let py = schema.py();
+        let discriminator = Discriminator::new(py, &schema.get_as_req(intern!(py, "discriminator"))?)?;
+
+        let choice_list: Bound<PyDict> = schema.get_as_req(intern!(py, "choices"))?;
+        let mut lookup: HashMap<String, CombinedSerializer> = HashMap::with_capacity(choice_list.len());
+        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(choice_list.len());
+
+        for (choice_key, choice_schema) in choice_list {
+            let serializer = CombinedSerializer::build(choice_schema.downcast()?, config, definitions).unwrap();
+            choices.push(serializer.clone());
+            lookup.insert(choice_key.to_string(), serializer);
+        }
+
+        let descr = choices
+            .iter()
+            .map(TypeSerializer::get_name)
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        Ok(Self {
+            discriminator,
+            lookup,
+            choices,
+            name: format!("TaggedUnion[{descr}]"),
+        }
+        .into())
+    }
+}
+
+impl_py_gc_traverse!(TaggedUnionSerializer { discriminator, lookup });
+
+impl TypeSerializer for TaggedUnionSerializer {
+    fn to_python(
+        &self,
+        value: &Bound<'_, PyAny>,
+        include: Option<&Bound<'_, PyAny>>,
+        exclude: Option<&Bound<'_, PyAny>>,
+        extra: &Extra,
+    ) -> PyResult<PyObject> {
+        let py = value.py();
+
+        let mut new_extra = extra.clone();
+        new_extra.check = SerCheck::Strict;
+
+        if let Some(tag) = self.get_discriminator_value(value) {
+            let tag_str = tag.to_string();
+            if let Some(serializer) = self.lookup.get(&tag_str) {
+                match serializer.to_python(value, include, exclude, &new_extra) {
+                    Ok(v) => return Ok(v),
+                    Err(err) => match err.is_instance_of::<PydanticSerializationUnexpectedValue>(py) {
+                        true => {
+                            if self.retry_with_lax_check() {
+                                new_extra.check = SerCheck::Lax;
+                                return serializer.to_python(value, include, exclude, &new_extra);
+                            }
+                        }
+                        false => return Err(err),
+                    },
+                }
+            }
+        }
+
+        let basic_union_ser = UnionSerializer::from_choices(self.choices.clone());
```

Comment:
Cloning here is potentially expensive, maybe store a `UnionSerializer` within the `TaggedUnionSerializer` so that fallback to that is cheap (and guaranteed not to fail)?